### ISSUE TITLE
Region.getSelf()

### ISSUE THREAD

**Issue by fcr** [htm-community/htm.core#610]
I'm in the middle of porting selected Python parts of htmresearch to htm.core and Python 3.
I can add new Python region classes, but without the `Region.getSelf()` method in the Python bindings I cannot access particular methods implemented by these new region classes.

The role of the`getSelf()` method is to cast the region handle back to the original Python class and allow the Python interpreter to access all methods and attributes of that class.

**Comment by dkeeney** [comment]
All C++ functions on the Region class are already exposed to python via pybind11.
See https://github.com/htm-community/htm.core/blob/master/bindings/py/cpp_src/bindings/engine/py_Engine.cpp

Which methods are you unable to access?

**Comment by fcr** [comment]
>>Which methods are you unable to access?
The ones in the Python class that extends the C++ region and that are not implemented by the C++ region.

I have attached an example class. For example, without `getSelf` to cast to the class RawSensor 
[RawSensor.py.zip](https://github.com/htm-community/htm.core/files/3473560/RawSensor.py.zip)

the method `RawSensor.addDataToQueue()' cannot be seen by the interpreter

**Comment by dkeeney** [comment]
Oh, you are talking about building a new Region Implementation for 'RawSensor' in python.  This is cool.

Now let me see if I understand what your problem is...
I think what you want to do is 
- create an Network object
- register your new Region implementation with the Network object as a new Region type.
- create a new region `r1` using that new registered region type.
- call `r1.getSelf().addDataToQueue(...)` on that region to populate the queue.  
- call Run to execute everything.

So what you want me to do is add this getSelf( ) function to the Region class which would return the instance of the RawSensor object that corresponds to it so that you can call addDataToQueue( ), and any other function on it.

This sounds like a reasonable thing to need.  Kind of a `getImplementation( )` function.  This would be useful for any of the sensor implementations.

I have to think about this.  If this were all python it would not be a problem.   The Region class is a generic stand-in for the implementation.  The caller could call it from either Python or from C++.
The implementation may have been implemented using C++ or in Python (or eventually it might be in any language that we might have bindings for).   

  In the general case it would have to be able to handle the 4 situations:
- Python caller, implementation in Python.   returns a python object.
- Python caller, implementation in C++.       returns the py bindings for the implementation.
- C++ caller, implementation in Python.      returns a pointer to some sort of wrapper class ????
- C++ caller, implementation in C++.    returns a pointer to a RegionImpl class that can be cast.
Doing this such that any function can be called would be difficult.

Here is a thought.  There is already a facility to pass a parameter to the Region class and it passes it on to the implementation.  There is one of these for each data type and it works in all four of the above situations.  There is also one for passing a named array of data setParameterArray(name, data) ... although I see the bindings are marked as TODO:     I suspect it is because the bindings for the Array object are not implemented.  The Array object can contain SDR objects and we do have bindings for that.

If we completed the bindings for this setParameterArray() function would it satisfy your requirements?  It does not allow calling any function but it could be used to implement a queue of array objects.

**Comment by fcr** [comment]
The solution using setParameterArray() would require an ugly change to the RawSensor since the addDataToQueue() takes 3 parameters that are synchronized, but it would work. However there are probably many situations that it would not.

If htm.core could just handle the first situation you listed, (Python caller, implementation in Python. returns a python object) that will go a long way to cover a lot of use cases. A search of the senders of Region.getSelf() in the nupic and htmresearch Python code base, shows that it is very heavily used. Having it in htm.core will make life much easier for people (like me) that want to port elements from that code to work with htm.core.

**Comment by dkeeney** [comment]
To be honest, this is the first time I have become aware of getSelf().   It is not documented as part of the official API and not implemented in any C++ code.  So the core library does not know about it.

It is a hack in the since that by using this you break all other languages, including C++ implemented apps.  I am disappointed that it seems to be used so much in nupic and htmresearch because we do want to port that code to our repository. 

The down-side is that if your code turns out to be useful to the general community (and I hope it will) then we will probably want to port it to C++ and include it into the library as a built-in region.  But with a hack like that we cannot do that.  It would be a python implemented Region usable only by Python apps.

I would think that we could come up with a better way of handling this.

I am leaning toward not doing this unless there is a consensus in the community to allow this.
@breznak, @ctrl-z-9000-times, @rhyolight , anyone else; what say ye?

**Comment by fcr** [comment]
Okay then, if the setParameterArray() will be working, I'll try and push it as far as I can.

I have a Python3 port of the ColumnPooler running on top off htm.core and it even passes all of its unit and integration tests. :-)

The RawSensor region was my current task, but I will wait for the implementation of setParameterArray() before continuing.

Meanwhile I will port the Python implementation of ApicalTiebreakTemporalMemory and its kin. (The C++ version would take me too long).

**Comment by dkeeney** [comment]
Ok, I will bump this up on my priority and try to get setParameterArray( ) binding working as soon as I can.   I would still like to hear what the community has to say about this issue.

**Comment by breznak** [comment]
>> I am leaning toward not doing this unless there is a consensus in the community to allow this.
> Okay then, if the setParameterArray() will be working, I'll try and push it as far as I can. 

I'm really glad you've found a working compromise for both! 

Just help me recap (as I'm catching up with last few days): 
- @fcr is porting some `htmresearch` code to be rebased on our `htm.core` (change from `nupic.core`)
  - that is really cool :+1: :+1: , please even consider including some of the (more stable, more used) classes as part of this repo for even more exposure. (We're not so "strict" in changes, so it's fine if it's not 100% tested (if mentioned!) and may be subject to later removal/rewrite..) I'd suggest just a `htm::research` namespace for these things to signal the possible API-breakage. 
- the `htm.research` as of now uses some `py.getSelf()` which is not part of the `NetworkAPI`
  - that would be fine, we could even extend the API
  - but @dkeeney says it is a hack that breaks all languages but python
    - if there's any other way, I'd also be agains such hack and require a proper (=working in c++,py) approach
    - if not, we could add it again in `htm::research` and tell users it'd kill their kitten if they ever use it! 
- you've agreed to try build on the ` setParameterArray() ` which is compatible and avoids the hack :+1:

**Comment by breznak** [comment]
PS: marking "won't fix" to indicate for now we want to go the other route, the discussion is still open.

**Comment by fcr** [comment]
Just to be clear I am currently porting just parts of the htmresearch Python code.
I have called the package "htm_advanced" rather than "htm_research" since it is no longer part of the research effort.
If anyone thinks that "htm_research" is a better name I have no problems changing.

**Comment by breznak** [comment]
> I have called the package "htm_advanced" rather than "htm_research" since it is no longer part of the research effort.

out of interest, what parts are you porting? You've mentioned ApicalTiebreakTM, ... 
and how "actively researched" vs stabilised & useful those are? I'm hinting in the direction if it'd mae sense to include them directly here?

**Comment by fcr** [comment]
I am basically aiming to get the location framework running on top of htm.core. This defines what parts I need to port. I have been using the location framework extensively and have found it stable and very useful to me.

If by "directly here" you mean htm.core, I would think that htmresearch.core would be more in line with what you are doing. But if you have a way to include Python code into htm.core I will be happy to contribute what I am doing - once I have it working. 

In any case, I planned to eventually put it on the Numenta community site.

Since I am building on the Connections class, I don't expect the Python code to have a performance problem. But if someone in the future wants to replace the code with a C++ version so much the better.

**Comment by breznak** [comment]
> If by "directly here" you mean htm.core, I would think that htmresearch.core would be more in line with what you are doing

yes, in htm.core, we aim to have both "production classes" (SP, TM) as well as research (Connections, there's GridCells encoder (only in py, too) ) 

> But if you have a way to include Python code into htm.core I will be happy to contribute what I am doing - once I have it working.

this sounds good. We'll discuss with you the experimental/usable ratio. But we do have a set-up way for python-only code, see `/py/htm/` which runs tests, builds and can use bindings nicely. 

>  location framework running

@ctrl-z-9000-times write the GridCell encoder in py here. 

>  building on the Connections class, I don't expect the Python code to have a performance problem.

(the c++ is not-only for performance, but some people prefer to/have frameworks in c++ only, that's why we have it as base). 
I wasn't talking that much about performance, more of API-breaks in Connections (consider it an internal class). This ties to the point above, if the code is useful (in generic point of view) to be included in here, we'd then have to fix the api-breaks, if it stays out of tree, it's upto the maintainer to resolve compatibility.

**Comment by dkeeney** [comment]
> PS: marking "won't fix" 

@breznak  Actually we do have to fix this because the binding for setParameterArray() was marked as TODO:   I am working on it now. ... actually I may need some help if this gets too complicated.

Note that this "fix" will address this one problem that @fcr had with code he is porting but not the problem of py.getSelf( ) in the general.   I have not looked at the other cases where py.getSelf( ) is used.

Basically this is a problem with keeping the plugins compatible across languages (each Region implementation being a plugin).  I hope my explanation above is understandable.

**Comment by fcr** [comment]
Thanks [@breznak](https://github.com/breznak.).

**Comment by dkeeney** [comment]
> RawSensor now works. :-)
Cool.

I have been giving a lot more thought to the more general problem of how to call arbitrary functions on the Region implementation and how to keep the calls synchronized with the execution.  Although I still don't have an ideal solution to replace getSelf(), I have thought of some additional tools that might help.

**executeCommand( )**
This is a way to invoke the Region implementation without breaking the plugin.  On the Region is this entry point:
```
/**
   * Request the underlying region to execute a command.
   * @param args
   *        A list of strings that the actual region will interpret.
   *        The first string is the command name. The other arguments are
   * optional.
   *
   * @returns
   *        The result value of command execution is a string determined
   *          by the underlying region.
   */
  virtual std::string executeCommand(const std::vector<std::string> &args);
```
If the region implementation chooses to implement it, there will be a function with this same signature (there is an additional index argument which I think is the iteration count).  Any call to this function on the Region object will be passed through to the Region implementation with the appropriate mapping of arguments for the language.  All of the arguments are strings so passing an array still requires setParameterArray( ).

**Callbacks**
There is still the problem of synchronization.  You want to feed the Region Implementation some data for each execute cycle such that everything is stopped while you do this.   The Network object in C++ does have a facility to perform a callback at the end of each run cycle.   The app can register a C function to be called at that point.  It is used to run the Watcher.  I don't see this implemented in the Python bindings but if there is interest I think this would be possible to do.

**VectorFileSensor**
This is a facility to read a file and then pass data, one array (or record) at a time, at each run cycle.  It is implemented as a separate region.  To feed data to an encoder you would first need for the encoder to have an input as well as an output.    To configure the network, create the VectorFileSensor region and your encoder region.  The setup a Link from the VectorFileSensor region to your encoder.  At each cycle the VectorFileSensor will read a record and output it on the link.  Your encoder could see this input and do its encoding of that data and perform its output.

I hope this gives you some ideas of how you might be useful in your design.
